### DISCUSSION TITLE
How can I make widget (and its descendants) insensitive to both mouse and keyboard?

### DISCUSSION THREAD

**Discussion by crftwr** [Textualize/textual#1515] [Q&A] [answered]
I would like to dynamically disable widget and its descendants and make insensitive to mouse and keyboard events. Can I do it with Textual APIs?

I know I can use can_focus / can_focus_children to make widgets not-focusable, so I can make them insensitive to keyboard events, but those widgets still get mouse inputs.

**Comment by davep** [comment, marked as answer]
Normally you'd do that with a widget's `disabled` property. See [this bit of code in one of Textual's examples](https://github.com/Textualize/textual/blob/779b10a0e8b8581fab512eb684a06eb90381705d/examples/five_by_five.py#L192-L199) -- it's used to disable the game while the "you've won" message is displayed (and of course later to reenable the game when a new game starts).

**Comment by crftwr** [comment]
@davep ,
I guess Button widget has "disabled" property, but other Widget types don't.

In my case, I would like to disable custom widgets (composite of Container and Static), and disable not only clicks but also while rotation, etc. Is there solution for that use-case?

**Reply by davep** [reply]
> I guess Button widget has "disabled" property, but other Widget types don't.

Doh! You're right of course. Chalk that up to only been one coffee into the day. 🤪

If you're rolling your own, perhaps take a glance at [how `Button` implements `disabled`](https://github.com/Textualize/textual/blob/main/src/textual/widgets/_button.py) and follow its lead?

**Reply by crftwr** [reply]
@davep thank you for your quick responses.

I understood that Button widget ignores click event by explicitly checking the "disabled" property in the event handler. It is definitely doable in my custom widget.

One hesitation to this approach was, when the custom widget is more complex than Button, and have many event handlers and actions, I may have to insert such "if self.disabled" checks in those multiple places within the widget.
But I am starting to think that it is the most straightforward way, because the meaning of "disabled" varies widget by widget.

Thanks!
